NFC: Add `fprint` variants of internal print functions

topolarity opened the pull request:
This adds/changes a number of `fprint_*` variants to internal debug functions:
  - rename + new `ios_t *` argument:
    - `jl_critical_error` -> `jl_fprint_critical_error`
    - `jl_print_native_codeloc` -> `jl_fprint_native_codeloc`
    - `jl_print_bt_entry_codeloc` -> `jl_fprint_bt_entry_codeloc`
    - `jl_gc_debug_print_status` -> `jl_gc_debug_fprint_status`
    - `jl_gc_debug_critical_error` -> `jl_gc_debug_fprint_critical_error`
 - new `fprint_*` variants:
    - `jl_safe_fprintf`
    - `jl_safe_static_show`
    - `jl_fprint_backtrace(t)`

Note that care needs to be taken w.r.t. what `ios_t *` value is passed into these functions from within a signal handler on Unix. Memory allocation is not async-signal-safe, so the ios_t needs to wrap a raw `fd_t` handle.

Immediate motivation is https://github.com/JuliaLang/julia/pull/54836

topolarity committed:
Implement `jl_safe_fprintf`

topolarity committed:
Add `fprint` variant of `jl_print_backtrace` and `jlbacktracet`

This allows backtraces to be printed somewhere other than stderr

topolarity committed:
Implement `jl_safe_static_show`

This is similar to `jl_static_show` but is wrapped in a setjmp() for
error tolerance.

topolarity committed:
Add `ios_t *` argument to jl_gc_debug_print_status and jl_gc_debug_critical_error

topolarity committed:
Replace `jl_show_sigill` with `jl_fprint_sigill`

topolarity committed:
Replace `jl_critical_error` with `jl_fprint_critical_error`

vtjnash commented:
I started to try to rebase this, but then noticed that it changed `jl_safe_printf` (to STDERR_FILENO) into `jl_unsafe_printf` (to ios_stderr), which is very much not NFC.

topolarity commented:
Is that not covered by:
> Note that care needs to be taken w.r.t. what ios_t * value is passed into these functions from within a signal handler on Unix.
> Memory allocation is not async-signal-safe, so the ios_t needs to wrap a raw fd_t handle.

vtjnash commented:
Ah, I guess that is true. But you're assuming that ios_t is async-signal-safe, which is only true in bm_none mode or if locally created. Do you want to rebase it then, but use `ios_t safe_stderr = ios_fd(&safe_stderr, STDERR_FILENO, 0, 0); safe_stderr->bm = bm_none; ios_close(&safe_stderr)`, where applicable to make it NFC?

topolarity committed:
Add `ios_safe_stderr` ios_t

This is not exposed to flisp and is intended for safe printing
in an async-signal context. Accordingly, it must be configured
to write to an `fd` with `bm_none`.

vtjnash commented on src/support/ios.c:
```diff
@@ -1074,6 +1076,17 @@ void ios_init_stdstreams(void)
     ios_fd(ios_stderr, STDERR_FILENO, 0, 0);
     ios_stderr->bm = bm_none;
+
+    // the following 'safe' variants must use `bm_line` / `bm_none`
```
`bm_line` seems likely very, very unsafe, so only `bm_none` is okay here. Only when allocated as a local variable, then bm_line is good and works to make something like stderr much faster without sacrificing safety.

topolarity replied:
Oh I understand your suggestion now. Happy to switch to the local variant with `bm_line` if you'd prefer (and will just change this to `bm_none` if not)

vtjnash replied:
I think this is fine, since changing to add the boilerplate everywhere seems unnecessary to add into this PR. If someone wants to do that later, they can and we'd remove `ios_safe_stderr` once done.

vtjnash commented on src/support/ios.c:
```diff
@@ -1061,6 +1061,8 @@ ios_t *ios_fd(ios_t *s, long fd, int isfile, int own)
 ios_t *ios_stdin = NULL;
 ios_t *ios_stdout = NULL;
 ios_t *ios_stderr = NULL;
+ios_t *ios_safe_stdout = NULL;
```
looks unused, please remove
